crates/oxc_semantic/src/jsdoc/builder.rs

```diff
@@ -0,0 +1,140 @@
+use std::{collections::BTreeMap, rc::Rc};
+
+use oxc_ast::{AstKind, GetSpan, Span, Trivias};
+
+use super::{JSDoc, JSDocComment};
+
+pub struct JSDocBuilder<'a> {
+    source_text: &'a str,
+
+    trivias: Rc<Trivias>,
+
+    docs: BTreeMap<Span, JSDocComment<'a>>,
+}
+
+impl<'a> JSDocBuilder<'a> {
+    pub fn new(source_text: &'a str, trivias: &Rc<Trivias>) -> Self {
+        Self { source_text, trivias: Rc::clone(trivias), docs: BTreeMap::default() }
+    }
+
+    pub fn build(self) -> JSDoc<'a> {
+        JSDoc::new(self.docs)
+    }
+
+    /// Save the span if the given kind has a jsdoc comment attached
+    pub fn retrieve_jsdoc_comment(&mut self, kind: AstKind<'a>) -> bool {
+        if !kind.is_declaration() {
+            return false;
+        }
+        let span = kind.span();
+        let comment_text = self.find_jsdoc_comment(span);
+        if let Some(comment_text) = comment_text {
+            self.docs.insert(span, JSDocComment::new(comment_text));
+        }
+        comment_text.is_some()
+    }
+
+    /// Find the jsdoc doc in frontend this span, a.k.a leading comment
+    fn find_jsdoc_comment(&self, span: Span) -> Option<&'a str> {
+        let (start, comment) = self.trivias.comments().range(..span.start).next()?;
+
+        if comment.is_single_line() {
+            return None;
+        }
+
+        let comment_text = Span::new(*start, comment.end()).source_text(self.source_text);
+
+        // Comments beginning with /*, /***, or more than 3 stars will be ignored.
+        let mut chars = comment_text.chars();
+        if chars.next() != Some('*') {
+            return None;
+        }
+        if chars.next() == Some('*') {
+            return None;
+        }
+
+        // The comment is the leading comment of this span if there is nothing in between.
+        let end = comment.end() + if comment.is_multi_line() { 2 } else { 0 };
+        let text_between = Span::new(end, span.start).source_text(self.source_text);
+        if !text_between.chars().all(char::is_whitespace) {
```

Comment:
I'm struggling to follow this logic – it feels like we should bail early if it _is_ all whitespace instead of not? Is this case covered in the tests?

Reply:
```
/** test */ ; function foo() {}
            ^ not a space so this jsdoc is not part of this function
```

Reply:
I changed it to `text_between.chars().any(|c| !c.is_whitespace())` to make it more readable.